### PR TITLE
Interactive transition

### PR DESCRIPTION
**Pull request by mattThousand**
TLDR: `SplitTransition` can still be used as before, but now it also offers support for interactive transitions.

Implemented:
- Added a `.Interactive` case to `TransitionType`. If the transition type is interactive, we allow user interaction during the transition
- Added a public `splitOffset` var, which, if set, causes the interactive transition to animate a spring out to the distance specified by `splitOffset` when the table view cell is first tapped. `splitOffset` is unused if `transitionType` is not `.Interactive`
- Added an internal `gestureRecognizer` var to store the pan gesture recognizer for the interactive transition. I'm storing it on the instance because it's used in multiple methods and I don't like passing it as a parameter each time. Thoughts on storing it vs passing it in willy nilly?
- Now storing the `fromViewController`, `toViewController`, `transitionContext`, and `container` view as properties as well, for the same reason
- Updated the example project

dec 19 2015 20 19

### CODE REVIEW

File: Shift-Demo/Shift-Demo.xcodeproj/xcuserdata/matthewbuckley.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist [hunk 1]
```diff
@@ -2,4 +2,22 @@
 <Bucket
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo.xcodeproj/xcuserdata/matthewbuckley.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:2]: You probably don't want to commit your `xcuserdata`.

File: Shift-Demo/Shift-Demo/DestinationViewController.swift [hunk 2]
```diff
@@ -13,10 +13,12 @@ class DestinationViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
 
-        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 7), dispatch_get_main_queue()) { () -> Void in
-            self.navigationController?.popViewControllerAnimated(true)
-        }
+        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: Selector("didTap:"))
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo/DestinationViewController.swift]: You can just put `"didTap:` here, strings are automatically converted to selectors.

File: Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+//  Shift-Demo
+//
+//  Created by Matthew Buckley on 12/19/15.
+//  Copyright © 2015 Raizlabs. All rights reserved.
+//
+
+import UIKit
+import Shift
+
+class SplitTransitionAnimatedViewController: UITableViewController {
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift]: As a rule, I declare all classes as `final` since the performance benefits of static dispatch come for free. When I need to subclass, I remove the `final` on the base class.

File: Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        initialNavigationControllerDelegate = navigationController?.delegate
+    }
+
+    // MARK: UITableViewDataSource
+
+    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
+        return 1
+    }
+    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 10
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift]: `colors.count()`

File: Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCellWithIdentifier("com.testApp.reuseIdentifier", forIndexPath: indexPath)
+        cell.selectionStyle = .None
+        cell.backgroundColor = colors[indexPath.row]
+        return cell
+    }
+
+    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
+        return 100.0
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift]: It's more performant to set `tableView.rowHeight` when all the cells have the same height.

File: Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+extension SplitTransitionAnimatedViewController: UINavigationControllerDelegate {
+
+    func navigationController(navigationController: UINavigationController,
+        animationControllerForOperation operation: UINavigationControllerOperation,
+        fromViewController fromVC: UIViewController,
+        toViewController toVC: UIViewController) -> UIViewControllerAnimatedTransitioning? {
+
+            if (operation == .Push && fromVC == self) {
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift:77]: You're getting `SwiftLint` warnings on these superfluous parentheses.

File: Shift-Demo/Shift-Demo/SplitTransitionInteractiveViewController.swift [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+//  Shift-Demo
+//
+//  Created by Matthew Buckley on 12/19/15.
+//  Copyright © 2015 Raizlabs. All rights reserved.
+//
+
+import UIKit
+import Shift
+
+class SplitTransitionInteractiveViewController: UITableViewController {
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo/SplitTransitionInteractiveViewController.swift]: Everything I said above applies here too :)

File: Shift-Demo/Shift-Demo/ViewController.swift [hunk 8]
```diff
@@ -38,51 +36,26 @@ class ViewController: UITableViewController {
...
-        cell.selectionStyle = .None
-        cell.backgroundColor = colors[indexPath.row]
+        cell.textLabel?.text = titles[indexPath.row]
+        cell.accessoryType = .DisclosureIndicator
         return cell
     }
 
     override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
-        return 100.0
+        return 50.0
```
Comment (jwatson) [inline Shift-Demo/Shift-Demo/ViewController.swift]: Use `tableView.rowHeight` instead of this delegate method.

File: Source/Shift/Classes/SplitTransition.swift [hunk 9]
```diff
@@ -88,19 +102,132 @@ public class SplitTransition: NSObject {
     }()
 
     /**
+     * In an interactive transition, splitOffset
+     * determines the vertical distance of the initial
+     * split when the cell is first tapped
+     */
+    public var splitOffset: CGFloat = 0.0
+
+    /**
```
Comment (jwatson) [inline Source/Shift/Classes/SplitTransition.swift]: Now I'm just getting nerdy, so feel free to ignore me. Single line comments can use the `///` syntax, and multi-line comments look like this now:

``` swift
/**
This is a super cool thing.

- Important: do not taunt the happy fun ball.
*/
```

File: Source/Shift/Classes/SplitTransition.swift [hunk 10]
```diff
@@ -88,19 +102,132 @@ public class SplitTransition: NSObject {
...
+    * Scroll distance in UI points
+    */
+    var interactiveTransitionScrollDistance: CGFloat = 0.0
```
Comment (jwatson) [inline Source/Shift/Classes/SplitTransition.swift]: I prefer to lean on the type inference system and use constructor syntax in these cases:

``` swift
var interactiveTransitionScrollDistance = CGFloat(0.0)
```

Good use of the `0.0` syntax though, I find that things can get confusing if you aren't a little specific about your literals.

File: Source/Shift/Classes/SplitTransition.swift [hunk 11]
```diff
@@ -88,19 +102,132 @@ public class SplitTransition: NSObject {
...
+            bottomSplitImageView.transform = CGAffineTransformMakeTranslation(0.0, (bottomPercentComplete * (bottomSplitImageView.bounds.size.height)))
+        }
+    }
+
+    /**
+    * For interactive transition, reflects whether
+    * the current pan is the first one (necessary to correctly
+    * manage transition progress vis a vis initial offset)
+    */
+    private var initialPan: Bool = true
```
Comment (jwatson) [inline Source/Shift/Classes/SplitTransition.swift]: You can get rid of the type declaration here.

**Comment by jwatson** [comment]
There's a `SwiftLint` warning about file line length violation that you might want to just ignore in the `.swiftlint.yml` file... Though you could refactor if you think it's worth it.

File: Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift (hunk 6 above)
Comment (mattThousand) [inline Shift-Demo/Shift-Demo/SplitTransitionAnimatedViewController.swift:77]: yikes, turns out I had an old install of `SwiftLint` that I had tweaked to suppress a few things I didn't like. Classic denial. working with the real deal now

File: Source/Shift/Classes/SplitTransition.swift (hunk 9 above)
Comment (mattThousand) [inline Source/Shift/Classes/SplitTransition.swift]: This is excellent advice. Do you use a plugin that generates the comment blocks for you or are you crafting your comments by hand?

**Comment by mattThousand** [comment]
Back to you, Tex

File: Source/Shift/Classes/SplitTransition.swift (hunk 9 above)
Comment (jwatson) [inline Source/Shift/Classes/SplitTransition.swift]: I use the mighty [VVDocumenter](https://github.com/onevcat/VVDocumenter-Xcode) plugin.

**Comment by jwatson** [comment]
Storing the gesture recognizer is reasonable; I tend to use parameter passing / dependency injection techniques for big external dependencies, not something that is somewhat integral to the object's job.

**Comment by mattThousand** [comment]
haha I'm not dogmatic about that sort of thing, yet. 'bend like the willow'

**Comment by jwatson** [comment]
Hah, 20+ years of K&R C have ruined me. Merging.
